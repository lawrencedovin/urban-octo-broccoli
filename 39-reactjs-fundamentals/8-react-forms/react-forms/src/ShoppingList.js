import React, { useState } from 'react';
import Item from './Item';
import NewItemForm from './NewItemForm';
import {v4 as uuid} from 'uuid';

const ShoppingList = () => {
    const INITIAL_STATE = [
        {id: uuid(), name: 'Peaut Butter', quantity: 2},
        {id: uuid(), name: 'Whole Milk', quantity: 1}
    ]
    const [items, setItems] = useState(INITIAL_STATE);
    const addItem = (newItem) => {
        // ...items saves previous items status and only adds new Item
        setItems(items => [...items, {id: uuid(), ...newItem}])
    }
    return (
        <div>
            <h3>Shopping List</h3>
            <NewItemForm addItem={addItem}/>
            <div>
                {items.map(({id, name, quantity}) => <Item id={id} name={name} quantity={quantity} key={id} />)}
            </div>
        </div>
    );
}

export default ShoppingList;