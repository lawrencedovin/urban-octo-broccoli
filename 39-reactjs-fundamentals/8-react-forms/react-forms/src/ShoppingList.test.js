import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ShoppingList from './ShoppingList';

it('renders without crashing', () => {
    render(<ShoppingList />);
});

it('matches snapshot', () => {
    const { asFragment } = render(<ShoppingList />);
    expect(asFragment()).toMatchSnapshot();
});

it('should add new item', () => {
    const { queryByText, getByLabelText } = render(<ShoppingList />);
    const input = getByLabelText('Name');
    const btn = queryByText('Add Item'); 
    expect(queryByText('Product: Chocolate Milk')).not.toBeInTheDocument();
    fireEvent.change(input, { target: { value: 'Chocolate Milk' } });
    fireEvent.click(btn);
    expect(queryByText('Product: Chocolate Milk')).toBeInTheDocument();
})