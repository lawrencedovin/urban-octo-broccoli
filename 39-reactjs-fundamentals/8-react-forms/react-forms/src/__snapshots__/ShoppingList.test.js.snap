// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Shopping List
    </h3>
    <form>
      <label
        for="name"
      >
        Name
      </label>
      <input
        id="name"
        name="name"
        placeholder="name"
        type="text"
        value=""
      />
      <label
        for="quantity"
      >
        Quantity
      </label>
      <input
        id="quantity"
        name="quantity"
        placeholder="quantity"
        type="number"
        value=""
      />
      <button>
        Add Item
      </button>
    </form>
    <div>
      <ul>
        <li>
          ID: e9ca1388-f12f-416b-b66a-9b96d6147a54
        </li>
        <li>
          Product: Peaut Butter
        </li>
        <li>
          Quantity: 2
        </li>
      </ul>
      <ul>
        <li>
          ID: 9db604f5-a70f-465f-ab7e-86265029ab9f
        </li>
        <li>
          Product: Whole Milk
        </li>
        <li>
          Quantity: 1
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
